[Quek Sze Long] iP

szelongq opened the pull request:
# Tipsy
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Tipsy frees your drunk mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~FAST~ *SUPER FAST* to use

All you need to do is,

1. download it from [here](https://github.com/szelongq/ip/releases).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

szelongq committed:
Implement Duke Level 1

Add greetings when starting Duke.
Add echoing of user input.
Add exiting of program on entering "bye".

szelongq committed:
Implement Duke Level 2

Add accepting and storing user inputted tasks into a list.
Add listing all tasks in the list.

szelongq committed:
Implement Duke Level 3

Add Task class.
Add ability to mark tasks as done.

szelongq committed:
Implement Duke Level-4

Add support for 3 types of tasks: ToDo, Deadline, Event.

szelongq committed:
Implement Automated Text UI Testing

szelongq committed:
Implement Duke Level 5

Add Exception checking for missing input for adding tasks.
Add Exception checking for unsupported operations.

szelongq committed:
Implement Duke Level 6

Add ability to delete tasks from the list.

szelongq committed:
Implement Enums for Task Types

szelongq committed:
Fix Bugs from Level 6

Fix Deadline tasks being created as Event tasks instead.
Update ToDo, Deadline and Event class constructors to set whether the task is done according to input.
Update Task Class to be abstract.

szelongq committed:
Implement Duke Level 7

Add ability to save and load tasks on the hard drive.

szelongq committed:
Implement Duke Level 8

Add ability to read and print dates for Deadline and Event tasks.

szelongq committed:
Merge branch 'branch-level-7'

szelongq committed:
Merge branch 'branch-Level-8'

szelongq committed:
Fix Bug from Merge with Level-7 and Level-8

Update the loading of saved tasks to parse dates.

szelongq committed:
Update Delete Feature to be work with Save Feature.

szelongq committed:
Add Ui, Storage and LoadingException Classes

Improve OOP.

szelongq committed:
Implement Duke A-MoreOOP

szelongq committed:
Implement A-Packages

Move all classes into duke package.

szelongq committed:
Implement Duke A-JUnit

Add JUnit Tests for TaskList and Parser classes.

szelongq committed:
Fix bug for when directory folder does not exist

szelongq committed:
Implement Duke A-JavaDoc

Add Javadoc comments for some of the classes.

szelongq committed:
Implement Duke A-CodingStandard

Edit some code to conform to CS2103T Coding Standard.

szelongq committed:
Implement Duke Level 9

Add ability to find tasks by searching with a keyword.

szelongq committed:
Merge branch 'branch-A-JavaDoc'

szelongq committed:
Merge branch 'branch-A-CodingStandard'

szelongq committed:
Merge branch 'branch-Level-9'

twothicc commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task{
```
Perhaps  adding a space after Task would be more consistent with your other class declarations.

szelongq replied:
Thanks, I have fixed it!

twothicc commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,71 @@
+    public int getSize() {
+        return tasks.size();
+    }
+
+    public Task getTask(int taskNum) {
```
Perhaps it would be good to add a header comment here to tell users that taskNum is using 1-based indexing.
This applies to other methods of TaskList too.

szelongq replied:
That's a great point, I have fixed it. Thank you!

twothicc commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,71 @@
+    /**
+     * Marks a task in the list as done.
+     *
+     * @param taskNum the number of the task to be marked as done.
```
Rather than number, wouldn't index be a more suitable term?

szelongq replied:
I agree! I have fixed it as well as for the other methods as well, thank you!

twothicc commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,50 @@
+import duke.*;
```
Perhaps it would be better to list imported classes explicitly. 

1. This could keep the import list minimal and up to date. 
2. The class would also be easier to comprehend and maintain.

szelongq replied:
Yes, I agree, I had missed this line when checking for code standard violations. Fixed it, thank you!

twothicc reviewed (commented):
LGTM :thumbsup:
Only some recommended changes related to coding standard.

cookiedan42 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,93 @@
+            } catch (DukeException e) {
+                ui.printErrorMessage(e.getMessage());
+            }
+        } while (true);
```
Consider making it obvious here what the loop exit condition is, either as a comment or by code structure. Without tracing executeCommand method, this initially looks like an infinite loop

szelongq replied:
Very true, I agree. I have added a comment to make clear what the loop exit condition is. Thank you!

cookiedan42 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,93 @@
+        case FIND_TASK:
+            findTask(Parser.parseSearchSubject(userInput));
+            break;
+        }
```
Consider explaining what happens when invalid input is passed, perhaps as a comment. Not immediately obvious that parseCommand throws exception and control will never flow to switch section

szelongq replied:
I agree. I have added more comments to elaborate what exceptions will be thrown when invalid inputs are entered. Thank you!

cookiedan42 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,110 @@
+        case "event":
+            taskType = TaskType.EVENT;
+            break;
+        }
```
Consider explicitly marking out the invalid case using default

szelongq replied:
Fixed it, thank you!

cookiedan42 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,104 @@
+                e.printStackTrace();
+                throw new LoadingException();
+            }
+            return null;
```
Do all callers of this function have to handle null case?
Consider returning empty ArrayList instead or documenting that the method can return null

szelongq replied:
That's a better alternative I had not thought of. I have updated it to return an empty ArrayList instead. Thank you!

szelongq committed:
Update Code as per Reviews.

Update code to follow coding standards better.
Add more comments to some methods for clarity of code function.

eezj35 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,143 @@
+    /**
+     * Prints the error message when the program encountered an error.
+     *
+     * @param message the error message from the exception.
```
Remember to capitalize the first letter in your description!

eezj35 commented on src/test/java/DukeTest.java:
```diff
@@ -0,0 +1,56 @@
+import duke.TaskList;
+import duke.ToDo;
+import duke.UnsupportedOperationException;
+import org.junit.jupiter.api.Test;
```
I think maybe you could group the duke package imports and the org.junit imports in separate groups?

eezj35 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,105 @@
+                writer.write(tasks.getTask(i).toSaveFormat() + "\n");
+            }
+            writer.close();
+        } catch(IOException ex){
```
remember the whitespace before the curly brace!

eezj35 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,143 @@
+     * @param subject the keyword that is being searched for.
+     */
+    public void printTasksWithSubject(TaskList tasks, String subject) {
+        int nResults = 0;
```
A very minor point, but maybe you could rename "nResults" to "numOfResults"? Just to improve overall readability

eezj35 commented on src/main/java/duke/CommandType.java:
```diff
@@ -0,0 +1,5 @@
+package duke;
+
+public enum CommandType {
+    EXIT, LIST, ADD_TASK, COMPLETE_TASK, DELETE_TASK, FIND_TASK
```
I like your use of enums for this case!

eezj35 reviewed (commented):
Great job on the code! It was actually quite hard for me to find much stuff to comment about, and I really liked that your methods were quite short and cleanly written!

szelongq committed:
Edit Code to Fit Coding Standard and Enhance Readability.

Capitalise first words of Javadoc descriptions.
Add spaces to follow coding standard.
Separate duke imports from org.junit imports.
Change variable name from nResults to numOfResults for better readability.

szelongq commented:
> Great job on the code! It was actually quite hard for me to find much stuff to comment about, and I really liked that your methods were quite short and cleanly written!

Thanks for the feedback! I have edited my code accordingly. :)

szelongq committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

szelongq committed:
Implement Duke A-Gradle

Use Gradle.

szelongq committed:
Merge branch 'branch-A-Gradle'

szelongq committed:
Implement A-CheckStyle

Rectify coding violations found by checkStyle.

szelongq committed:
Merge branch 'branch-A-CheckStyle'

szelongq committed:
Implement Duke Level 10

Add a GUI using JavaFX.
Remove text-based UI.

szelongq committed:
Merge branch 'branch-Level-10'

szelongq committed:
Fix bugs preventing JAR file generation.

szelongq committed:
Use Assertions

There are several methods that convert a collection to a set. In somerrr
There are a number of assumptions made in the code about the
program's flow of control and input of certain non-public methods.

These assumptions are not checked and could potentially lead to bugs
that are hidden in the program.

Let's use assertions to document these important assumptions that
should hold true during execution of the code.

Assertions do not have a heavy impact on performance and allows
quick detection of bugs arising from the assumptions made in the code.
They also help to document some of the logic in the code and improves
maintainability of the program.

szelongq committed:
Parser class: move parsing of TaskType to TaskType class

The parseNewTask method is considerably long.

Long methods are harder to read. Bugs are also harder to find in
long methods.

Let's move the parsing of TaskType from a String object into a new
method in TaskType class.

This would reduce the length of the parseNewTask method,
making it easier to read and understand. It would also allow
other future classes to use the method as well.

szelongq committed:
Storage class: create a new method for parsing a saved task

The loadTasks method is too long.

Let's move the parsing of a single saved task into a new loadTask method.

This would result in two shorter and more readable methods.

szelongq committed:
Improve code readability

Break up lines in DialogBox class.

Introduce more intermediate variables to improve readability and
explicitness of code in Parser and Storage classes.

szelongq committed:
Merge pull request #1 from szelongq/branch-A-Assertions

Use assertions in code

szelongq committed:
Merge branch 'master' into branch-A-CodeQuality

szelongq committed:
Merge pull request #2 from szelongq/branch-A-CodeQuality

Improve code quality

szelongq committed:
Create main.yml

Add new workflow to the project.

szelongq committed:
Add whitespaces to follow checkstyle format

szelongq committed:
Merge pull request #3 from szelongq/branch-A-CI

Set up workflow for continuous integration

szelongq committed:
Add ability for the user to change the target save file

Add two new commands: showpath and setpath.

szelongq committed:
Add javadoc comments

szelongq committed:
Merge pull request #4 from szelongq/branch-C-FlexibleDataSource

Add extension for flexible data source

szelongq committed:
Improve GUI

Change display pictures for Duke and User.
Add product name in the title bar of the GUI window.
Add bordering and background to messages.

szelongq committed:
Merge pull request #6 from szelongq/branch-A-BetterGui

Improve GUI

szelongq committed:
Rebrand Duke as Tipsy

szelongq committed:
Fix Checkstyle Errors

szelongq committed:
Merge pull request #7 from szelongq/branch-Rebranding

Rebrand Duke as Tipsy

szelongq committed:
Update README.md

Rebrand Duke as Tipsy

szelongq committed:
Add Product Screenshot

szelongq committed:
Update User Guide

szelongq committed:
Change to using gradle v6.2 instead of v6.3
